### PR TITLE
CAMEL-9591 - Support Saxon integrated extension functions

### PR DESCRIPTION
**Pull request by lburgazzoli**

### CODE REVIEW

File: camel-core/src/main/java/org/apache/camel/component/xslt/XsltHelper.java [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+                    Class<?> extensionClass = camelContext.getClassResolver().resolveMandatoryClass(
+                        SAXON_EXTENDED_FUNCTION_DEFINITION_CLASS_NAME, XsltComponent.class.getClassLoader()
+                    );
+
+                    method = configuration.getClass().getMethod("registerExtensionFunction", extensionClass);
+                    if (method != null) {
+                        factory.setFeature(XMLConstants.FEATURE_SECURE_PROCESSING, true);
+                        for (Object extensionFunction : saxonExtensionFunctions) {
+                            if (extensionClass.isInstance(extensionFunction)) {
+                                LOG.debug("Sazon.registerExtensionFunction {}", extensionFunction);
```
Comment (davsclaus) [inline camel-core/src/main/java/org/apache/camel/component/xslt/XsltHelper.java]: there is a typo here

File: camel-core/src/main/java/org/apache/camel/component/xslt/XsltEndpoint.java [hunk 2]
```diff
@@ -71,6 +72,8 @@
     @UriParam
     private boolean saxon;
     @UriParam(label = "advanced")
+    private List<Object> saxonExtensionFunctions;
+    @UriParam(label = "advanced")
```
Comment (davsclaus) [inline camel-core/src/main/java/org/apache/camel/component/xslt/XsltEndpoint.java]: multivalue = "true"

File: camel-core/src/main/java/org/apache/camel/component/xslt/XsltEndpoint.java [hunk 3]
```diff
@@ -198,6 +201,18 @@ public void setSaxon(boolean saxon) {
         this.saxon = saxon;
     }
 
+    public List<Object> getSaxonExtensionFunctions() {
+        return saxonExtensionFunctions;
+    }
+
+    /**
+     * Allows you to use a custom net.sf.saxon.lib.ExtensionFunctionDefinition.
+     * You would need to add Saxon to the classpath.
```
Comment (davsclaus) [inline camel-core/src/main/java/org/apache/camel/component/xslt/XsltEndpoint.java]: add something that you can use one or more custom functions, eg its a list.

File: camel-core/src/main/java/org/apache/camel/component/xslt/XsltEndpoint.java [hunk 4]
```diff
@@ -71,6 +72,8 @@
     @UriParam
     private boolean saxon;
     @UriParam(label = "advanced")
+    private List<Object> saxonExtensionFunctions;
```
Comment (davsclaus) [inline camel-core/src/main/java/org/apache/camel/component/xslt/XsltEndpoint.java]: You need to add some logic to xstl component so people can provide the functions in the ur, as a comma separated value of names to lookup in the registry, eg `saxonExtensionFunctions=#foo,#bar` etc. to lookup foo and bar from the registry. eg add an unit test in camel-spring to see how easy/hard it is to use as a spring/blueprint end user.

**Comment by davsclaus** [comment]
Ah yeah you need to add unit tests to camel-saxon both for java and spring

File: camel-core/src/main/java/org/apache/camel/component/xslt/XsltEndpoint.java (hunk 4 above)
Comment (lburgazzoli) [inline camel-core/src/main/java/org/apache/camel/component/xslt/XsltEndpoint.java]: Is there any out-of-the box support in camel ? or I need to define saxonExtensionFunctions as String and then split/lookup values in the component ?

Comment (davsclaus) [inline camel-core/src/main/java/org/apache/camel/component/xslt/XsltEndpoint.java]: Yeah you would need to use org.apache.camel.util.EndpointHelper#resolveReferenceListParameter in the component. You can git grep to find other components using it. Then the type in the endpoint can be the same.

**Comment by lburgazzoli** [comment]
- fixed typo
- make use of resolveReferenceListParameter
- test case added

**Comment by davsclaus** [comment]
Thanks the code has been merged. Do you mind adding it to the docs at
http://camel.apache.org/xslt

And maybe add a little example how to use it?

**Comment by davsclaus** [comment]
And close this PR

**Comment by lburgazzoli** [comment]
Should the example be added to camel-saxon ?

**Comment by davsclaus** [comment]
no to xslt

**Comment by lburgazzoli** [comment]
Done

**Pull request closed without merging**